driver/src/driver/scheduler/system.rs

```diff
@@ -62,92 +65,131 @@ impl<'a> SystemScheduler<'a> {
-        let intended_solve_start_time = solving_batch.solve_start_time()
-            + self.auction_timing_configuration.target_start_solve_time;
-        // unwrap here because this cannot fail because the `solving_batch`'s
-        // start time is always before `now`.
-        let elapsed_time = now
-            .duration_since(solving_batch.solve_start_time())
+/// Return current batch id and what to do with it.
+fn determine_action(
+    auction_timing_configuration: AuctionTimingConfiguration,
+    last_solved_batch: Option<BatchId>,
+    now: SystemTime,
+) -> Result<Action> {
+    let solving_batch = BatchId::currently_being_solved(now)
+        .context("failed to get batch id currently being solved")?;
+    let intended_solve_start_time =
+        solving_batch.solve_start_time() + auction_timing_configuration.target_start_solve_time;
+    // unwrap here because this cannot fail because the `solving_batch`'s
+    // start time is always before `now`.
+    let elapsed_time = now
+        .duration_since(solving_batch.solve_start_time())
+        .unwrap();
+
+    let action = if last_solved_batch == Some(solving_batch)
+        || elapsed_time >= auction_timing_configuration.solver_time_limit
+    {
+        let next = solving_batch.next();
+        let duration = (next.solve_start_time()
+            + auction_timing_configuration.target_start_solve_time)
+            .duration_since(now)
             .unwrap();
+        Action::Sleep(duration)
+    } else if now < intended_solve_start_time {
+        let duration = intended_solve_start_time.duration_since(now).unwrap();
+        Action::Sleep(duration)
+    } else {
+        let time_limit = auction_timing_configuration.solver_time_limit - elapsed_time;
+        Action::Solve(solving_batch, time_limit)
+    };
+
+    Ok(action)
+}
 
-        let action = if self.last_solved_batch == Some(solving_batch)
-            || elapsed_time >= self.auction_timing_configuration.solver_time_limit
-        {
-            let next = solving_batch.next();
-            let duration = (next.solve_start_time()
-                + self.auction_timing_configuration.target_start_solve_time)
-                .duration_since(now)
-                .unwrap();
-            Action::Sleep(duration)
-        } else if now < intended_solve_start_time {
-            let duration = intended_solve_start_time.duration_since(now).unwrap();
-            Action::Sleep(duration)
-        } else {
-            let time_limit = self.auction_timing_configuration.solver_time_limit - elapsed_time;
-            Action::Solve(solving_batch, time_limit)
-        };
+fn should_attempt_to_solve_again(
+    auction_timing_configuration: AuctionTimingConfiguration,
+    now: SystemTime,
+    batch_id: BatchId,
+    driver_result: &DriverResult,
+) -> bool {
+    match driver_result {
+        DriverResult::Ok => false,
+        DriverResult::Retry(_) => match determine_action(auction_timing_configuration, None, now) {
```

Comment:
All we really want to know is if we still have time to solve batch_id again. While reusing `determine_action` happens to work, I found it hard to verify that (it has a lot more logic taking last_batch_id and the one that is currently being solved into account) and it might be simpler to just rewrite the logic in something like `can_still_solve_batch`. Maybe `determine_action` can call reuse that method.

Reply:
Agree. Will implement.